commons/zenoh-buffers/src/zslice.rs

```diff
@@ -30,17 +31,59 @@ use core::{
 /*************************************/
 pub trait ZSliceBuffer: Send + Sync + fmt::Debug {
     fn as_slice(&self) -> &[u8];
+
+    #[cfg(feature = "shared-memory")]
+    /// # Safety
+    /// Get mutable access to underlying data. This is safe if
+    /// there is no sharing of underlying data.
+    unsafe fn as_mut_slice_unchecked(&mut self) -> &mut [u8];
+    #[cfg(feature = "shared-memory")]
+    fn as_mut_slice(&mut self) -> Option<&mut [u8]>;
+
+    #[cfg(not(feature = "shared-memory"))]
     fn as_mut_slice(&mut self) -> &mut [u8];
+
     fn as_any(&self) -> &dyn Any;
+    #[cfg(feature = "shared-memory")]
+    fn is_valid(&self) -> bool {
+        true
+    }
+}
+
+/// # Safety
+/// Get mutable access to underlying data. For "shared-memory" feature
+/// this is safe if there is no sharing of underlying data for SHM buffers,
+/// for not "shared-memory" operation this is always safe. This function is
+/// intended to make ZSlice's interface more homogenious as it differs depending
+/// on "shared-memory" feature
+#[inline]
+pub unsafe fn as_mut_slice_featureless<T: ZSliceBuffer + ?Sized>(slice: &mut T) -> &mut [u8] {
+    #[cfg(feature = "shared-memory")]
+    return slice.as_mut_slice_unchecked();
+
+    #[cfg(not(feature = "shared-memory"))]
+    slice.as_mut_slice()
 }
 
 impl ZSliceBuffer for Vec<u8> {
     fn as_slice(&self) -> &[u8] {
         self.as_ref()
     }
+
+    #[cfg(feature = "shared-memory")]
+    unsafe fn as_mut_slice_unchecked(&mut self) -> &mut [u8] {
```

Comment:
I believe the concept of mutability in the slice requires a bit more of discussion, at least how it is exposed via different features.

Reply:
Please see [this answer](https://github.com/eclipse-zenoh/zenoh/pull/823#discussion_r1535429983)